clippy_lints/src/test_without_fail_case.rs

```diff
@@ -0,0 +1,268 @@
+use clippy_config::Conf;
+use clippy_utils::diagnostics::span_lint_and_note;
+use clippy_utils::macros::root_macro_call_first_node;
+use clippy_utils::ty::is_type_diagnostic_item;
+use clippy_utils::visitors::Visitable;
+use clippy_utils::{is_in_test_function, method_chain_args};
+use rustc_data_structures::fx::FxHashSet;
+use rustc_hir::def::Res;
+use rustc_hir::def_id::DefId;
+use rustc_hir::intravisit::{self, Visitor};
+use rustc_hir::{AnonConst, Expr, ExprKind, Item, ItemKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::hir::nested_filter;
+use rustc_middle::ty;
+use rustc_session::impl_lint_pass;
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for test functions that cannot fail.
+    ///
+    /// ### Why is this bad?
+    /// A test function that cannot fail might indicate that it does not actually test anything.
+    /// It could lead to false positives in test suites, giving a false sense of security.
+    ///
+    /// ### Example
+    /// ```rust
+    /// #[test]
+    /// fn my_test() {
+    ///     let x = 2;
+    ///     let y = 2;
+    ///     if x + y != 4 {
+    ///         eprintln!("this is not a correct test")
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// Use instead:
+    /// ```rust
+    /// #[test]
+    /// fn my_test() {
+    ///     let x = 2;
+    ///     let y = 2;
+    ///     assert_eq!(x + y, 4);
+    /// }
+    /// ```
+    #[clippy::version = "1.82.0"]
+    pub TEST_WITHOUT_FAIL_CASE,
+    restriction,
+    "test function cannot fail because it does not call anywhere to panic or assert"
+}
+
+pub struct TestWithoutFailCase {
+    config: SearchConfig,
+}
+
+impl TestWithoutFailCase {
+    pub fn new(conf: &Conf) -> Self {
+        Self {
+            config: SearchConfig {
+                indexing_fallible: conf.test_without_fail_case_include_indexing_as_fallible,
+                fallible_paths: conf.test_without_fail_case_fallible_paths.iter().cloned().collect(),
+                non_fallible_paths: conf.test_without_fail_case_non_fallible_paths.iter().cloned().collect(),
+            },
+        }
+    }
+}
+
+impl_lint_pass!(TestWithoutFailCase => [TEST_WITHOUT_FAIL_CASE]);
+
+impl<'tcx> LateLintPass<'tcx> for TestWithoutFailCase {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'tcx>) {
+        // Only interested in functions that are annotated with `#[test]`.
+        if let ItemKind::Fn(_, _, body_id) = item.kind
+            && is_in_test_function(cx.tcx, item.hir_id())
```

Comment:
This lint only runs on tests. This also means, that the lint can only detect something, if Clippy is run with `cargo clippy --all-targets` or `cargo clippy --tests`. This should be mentioned in the documentation of the lint.

I'm not sure, if `lintcheck` is run with `--all-targets`. It at least didn't trigger once for all the crates we test on.